src/EditorFeatures/CSharp/CommentSelection/ToggleBlockCommentCommandHandler.cs

```diff
@@ -0,0 +1,428 @@
+using Microsoft.VisualStudio.Text.Operations;
+using Microsoft.VisualStudio.Utilities;
+using Roslyn.Utilities;
+using VSCommanding = Microsoft.VisualStudio.Commanding;
+
+namespace Microsoft.CodeAnalysis.Editor.CSharp.CommentSelection
+{
+    /* TODO - Modify these once the toggle block comment handler is added.
+    [Export(typeof(VSCommanding.ICommandHandler))]
+    [ContentType(ContentTypeNames.RoslynContentType)]
+    [Name(PredefinedCommandHandlerNames.CommentSelection)]*/
+    internal class ToggleBlockCommentCommandHandler :
+        AbstractCommentSelectionCommandHandler/*,
+        VSCommanding.ICommandHandler<CommentSelectionCommandArgs>*/
+    {
+        [ImportingConstructor]
+        internal ToggleBlockCommentCommandHandler(
+            ITextUndoHistoryRegistry undoHistoryRegistry,
+            IEditorOperationsFactoryService editorOperationsFactoryService) : base(undoHistoryRegistry, editorOperationsFactoryService)
+        { }
+
+        /* TODO - modify once the toggle block comment handler is added.
+        public VSCommanding.CommandState GetCommandState(CommentSelectionCommandArgs args)
+        {
+            return GetCommandState(args.SubjectBuffer);
+        }
+
+        public bool ExecuteCommand(CommentSelectionCommandArgs args, CommandExecutionContext context)
+        {
+            return ExecuteCommand(args.TextView, args.SubjectBuffer, Operation.Toggle, context);
+        }*/
+
+        public override string DisplayName => EditorFeaturesResources.Toggle_Block_Comment;
+
+        protected override string GetTitle(Operation operation)
+        {
+            return EditorFeaturesResources.Toggle_Block_Comment;
+        }
+
+        protected override string GetMessage(Operation operation)
+        {
+            return EditorFeaturesResources.Toggling_block_comment_on_selection;
+        }
+
+        protected override void SetTrackingSpans(ITextView textView, ITextBuffer buffer, List<CommentTrackingSpan> trackingSpans)
+        {
+            var spans = trackingSpans.Select(trackingSpan => trackingSpan.ToSelection(buffer));
+            textView.GetMultiSelectionBroker().SetSelectionRange(spans, spans.Last());
+        }
+
+        internal override void CollectEdits(Document document, ICommentSelectionService service, NormalizedSnapshotSpanCollection selectedSpans,
+            List<TextChange> textChanges, List<CommentTrackingSpan> trackingSpans, Operation operation, CancellationToken cancellationToken)
+        {
+            var experimentationService = document.Project.Solution.Workspace.Services.GetRequiredService<IExperimentationService>();
+            if (!experimentationService.IsExperimentEnabled(WellKnownExperimentNames.RoslynToggleBlockComment))
+            {
+                return;
+            }
+
+            if (selectedSpans.IsEmpty())
+            {
+                return;
+            }
+
+            var getRoot = document.GetSyntaxRootAsync();
```

Comment:
1. pass cancellation token. 
2. why is the containing method method not async?
3. if this must be synchronous, you should call GetSyntaxRootSynchronously.

Reply:
As far as I am aware, there is not an async command handler (correct me if I'm wrong).  This method could be async, but I'd just be moving the wait up a method call
1/3 will do.

Reply:
> As far as I am aware, there is not an async command handler (correct me if I'm wrong). This method could be async, but I'd just be moving the wait up a method call
1/3 will do.

Yes, tha'ts appropriate.  We try to make it all async underneath, and we do the one final wait all the way up at the actual command handler entrypoint.  Thanks!